contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol

```diff
@@ -163,27 +149,74 @@ contract LinkAvailableBalanceMonitor is ConfirmedOwner, Pausable, AutomationComp
-    uint256 stopIdx = targetAddresses.length;
-    uint256 numCanFund = LINK_TOKEN.balanceOf(address(this)) / topUpAmount;
-    stopIdx = numCanFund < stopIdx ? numCanFund : stopIdx;
-    for (uint256 idx = 0; idx < stopIdx; idx++) {
-      (bool exists, uint256 minBalance) = s_watchList.tryGet(targetAddresses[idx]);
-      if (!exists) {
-        emit TopUpBlocked(targetAddresses[idx]);
-        continue;
+    MonitoredAddress memory target;
+    uint256 localBalance = i_linkToken.balanceOf(address(this));
+    address[] memory targetsTopUp = new address[](targetAddresses.length);
+    for (uint256 idx = 0; idx < targetAddresses.length; idx++) {
+      address targetAddress = targetAddresses[idx];
+      target = s_targets[targetAddress];
+      (bool needsFunding, ) = _needsFunding(targetAddress, target.minBalance);
+      if (
+        target.topUpAmount > 0 && localBalance >= target.topUpAmount && !(targetAddress == address(0)) && needsFunding
+      ) {
+        targetsTopUp[idx] = targetAddress;
+        localBalance -= target.topUpAmount;
+      } else {
+        emit TopUpBlocked(targetAddress);
       }
-      (bool needsFunding, address target) = _needsFunding(targetAddresses[idx], minBalance);
-      if (!needsFunding) {
-        emit TopUpBlocked(targetAddresses[idx]);
+    }
+    for (uint256 idx = 0; idx < targetsTopUp.length; idx++) {
+      address targetAddress = targetsTopUp[idx];
+      if (targetAddress == address(0)) {
         continue;
       }
-      LINK_TOKEN.transfer(target, topUpAmount);
-      emit TopUpSucceeded(targetAddresses[idx]);
+      target = s_targets[targetAddress];
+      bool success = i_linkToken.transfer(targetAddress, target.topUpAmount);
+      if (success) {
+        target.lastTopUpTimestamp = uint56(block.timestamp);
+        emit TopUpSucceeded(targetAddress);
+      } else {
+        emit TopUpFailed(targetAddress);
+      }
+    }
+  }
+
+  /// @notice checks the target (could be direct target or IAggregatorProxy), and determines
+  /// if it is elligible for funding
+  /// @param targetAddress the target to check
+  /// @param minBalance minimum balance required for the target
+  /// @return bool whether the target needs funding or not
+  /// @return address the address of the contract needing funding
+  function _needsFunding(address targetAddress, uint256 minBalance) private view returns (bool, address) {
+    uint256 minWaitPeriodSeconds = s_minWaitPeriodSeconds;
+    MonitoredAddress memory addressToCheck = s_targets[targetAddress];
+    ILinkAvailable target;
+    IAggregatorProxy proxy = IAggregatorProxy(targetAddress);
+    // Explicitly check if the targetAddress is the zero address
+    // or if it's not a contract. In both cases return with false,
+    // to prevent target.linkAvailableForPayment from running,
+    // which would revert the operation.
+    if (targetAddress == address(0) || !(targetAddress.code.length > 0)) {
+      return (false, address(0));
+    }
```

Comment:
can we make sure there is a test that covers this case? IIRC, this is the case that is failing in production right now. I would love to see a before/after test that show this case failing before making contract changes, and passing afterwards.

Reply:
Test created!